File: test/conformance/js/rows/R16.mjs
// R16 — §5.4's counters exactly: unknown once per peer at COMPILE and never
// per record; widened once per entry per record, a folded element run is ONE;
// clamped once per entry per record for count/text; the bounds pass counts a
// forged ordinal remapped to None on BOTH plans; copy/const/present/ordinal
// move nothing.
//
// Production path: CellFixedLoad → TableFixedRun → each op code. Every counter
// assertion hits the generated JS emitter (build/tables-generated-js/v1/Tblv1Table.js).
//
// Plan format: 9 int32 lanes per entry — [op,src,dst,size,aux,guard,arg,meta,argw]
// Ops: Copy=0, Count=1, Ordinal=3, Widen=4, Const=5, WidenF=6
// Guard=-1 means no guard (always fires); unset default is 0 which may block.
// All entries below set guard to -1 so every op runs.

import { readFileSync } from "node:fs";
import { join, resolve } from "node:path";
import { pathToFileURL } from "node:url";

const generated = process.env.SCHEMA_JS_GENERATED ?? "build/tables-generated-js";
const base = resolve(generated);
const tbl = await import(pathToFileURL(join(base, "v1", "Tblv1Table.js")).href);

const {
  TableFixedReport,
  TableFixedResetReport,
  TableFixedRun,
  TableFixedPlan,
  TableFixedLanes,
} = tbl;

// Lanes matching Tblv1Table.js
const L = { Op:0, Src:1, Dst:2, Size:3, Aux:4, Guard:5, Arg:6, Meta:7, ArgW:8 };
// Ops matching Tblv1Table.js
const O = { Copy:0, Count:1, Text:2, Ordinal:3, Widen:4, Const:5, WidenF:6 };
const NoGuard = -1;

let failures = [];
function ok(cond, label) {
  if (!cond) { failures.push(label); console.log("FAIL: " + label); }
  else console.log("ok: " + label);
}

// Build one plan entry with NO GUARD (so it always fires)
function mk(op, src, dst, size, aux, arg, meta, argw) {
  const e = new Int32Array(TableFixedLanes);
  e[L.Op]   = op | 0;
  e[L.Src]  = src | 0;
  e[L.Dst]  = dst | 0;
  e[L.Size] = size | 0;
  e[L.Aux]  = aux | 0;
  e[L.Guard]= NoGuard; // always fire
  e[L.Arg]  = arg | 0;
  e[L.Meta] = meta | 0;
  e[L.ArgW] = argw | 0;
  return e;
}

// ---- TEST 1: copy moves nothing ----
function test_copy() {
  const report = new TableFixedReport();
  const src = new Uint8Array([0xab, 0xcd, 0xef]);
  const dst = new Uint8Array(3);
  const dv = new DataView(dst.buffer);
  TableFixedRun(mk(O.Copy, 0, 0, 3), 1, src, dv, 0, dst, dv, null, report);
  ok(report.unknown === 0, "copy: unknown === 0");
  ok(report.kindMismatch === 0, "copy: kindMismatch === 0");
  ok(report.widened === 0, "copy: widened === 0");
  ok(report.clamped === 0, "copy: clamped === 0");
}

// ---- TEST 2: widened once per entry per record ----
function test_widen() {
  const report = new TableFixedReport();
  const src = new Uint8Array([0xff]);      // int8 = -1 (sign bit set)
  const dst = new Uint8Array(4);           // int32 dest
  const dv = new DataView(dst.buffer);
  TableFixedRun(mk(O.Widen, 0, 0, 1, 0, 0, 4 | (1 << 8), 1),
    1, src, dv, 0, dst, dv, null, report);
  ok(report.widened === 1, "widen: widened === 1 (once per entry)");
  ok(report.clamped === 0, "widen: clamped === 0");
  ok(dv.getInt32(0, true) === -1, "widen: value sign-extended to -1");
}

// ---- TEST 3: clamped — count negative (high-bit set u32) ----
function test_count_clamp_negative() {
  const report = new TableFixedReport();
  // source[0..3] = 0xFFFFFFFF = -1 signed, source[4..7] = junk data
  const src = new Uint8Array([0xff, 0xff, 0xff, 0xff, 0xaa, 0xbb, 0xcc, 0xdd]);
  const dst = new Uint8Array(8);
  const dv = new DataView(dst.buffer);
  TableFixedRun(mk(O.Count, 0, 0, 4, 0, 0, 0, 1),
    1, src, dv, 0, dst, dv, null, report);
  ok(report.clamped >= 1, "count negative: clamped >= 1 (u32 > bound 4)");
  ok(dv.getUint32(0, true) <= 4, "count negative: value clamped to ≤ 4");
  ok(report.widened === 0, "count negative: widened === 0");
}

// ---- TEST 4: clamped — count above bound ----
function test_count_clamp_above() {
  const report = new TableFixedReport();
  const src = new Uint8Array([0x64, 0x00, 0x00, 0x00, 0xaa, 0xbb, 0xcc, 0xdd]);
  const dst = new Uint8Array(8);
  const dv = new DataView(dst.buffer);
  TableFixedRun(mk(O.Count, 0, 0, 2, 0, 0, 0, 1),
    1, src, dv, 0, dst, dv, null, report);
  ok(report.clamped === 1, "count above: clamped === 1 (100 > 2)");
  ok(dv.getUint32(0, true) === 2, "count above: clamped to 2");
}

// ---- TEST 5: text clamped ----
function test_text_clamp() {
  const report = new TableFixedReport();
  // source[0..3] = huge count, [4..5] = "Hi" payload
  const src = new Uint8Array([0xff, 0xff, 0xff, 0xff, 0x48, 0x69, 0xAA, 0xBB]);
  const dst = new Uint8Array(10);
  const dv = new DataView(dst.buffer);
  // Text: UTF8, cap=size=3, count=overflow → clamped to 3
  TableFixedRun(mk(O.Text, 0, 4, 3, 7, 0, 0, 1),
    1, src, dv, 0, dst, dv, null, report);
  ok(report.clamped === 1, "text clamp: clamped === 1");
}

// ---- TEST 6: forged ordinal clamped on compiled plan ----
function test_forged_ordinal_compiled() {
  const report = new TableFixedReport();
  const src = new Uint8Array([0x05, 0x00, 0x00, 0x00]);  // ordinal 5
  const dst = new Uint8Array(4);
  const dv = new DataView(dst.buffer);
  const remap = new Int32Array(4);
  remap[0] = 2; // writer extent = 2 (valid ordinals: 0,1,2)
  TableFixedRun(mk(O.Ordinal, 0, 0, 4, 0, 0, 0, 4),
    1, src, dv, 0, dst, dv, remap, report);
  ok(report.clamped === 1, "forged ordinal: clamped === 1 (5 > 2)");
  ok(dv.getUint32(0, true) === 0, "forged ordinal: remapped to None (0)");
  ok(report.widened === 0, "forged ordinal: widened === 0");
}

// ---- TEST 7: forged ordinal clamped on identity-equivalent path (Const) ----
// The unguarded Const entry reads ordinal bytes and checks raw > arms (meta).
// Same forged bytes land clamped == 1 on both paths (§5.4).
function test_forged_ordinal_const_path() {
  const reportCompiled = new TableFixedReport();
  const reportConst = new TableFixedReport();
  const src = new Uint8Array([0x05, 0x00, 0x00, 0x00, 0xaa, 0xbb, 0xcc, 0xdd]);
  const dstC = new Uint8Array(4);
  const dstA = new Uint8Array(4);
  const remapC = new Int32Array(4);
  remapC[0] = 2; // writer extent
  const revC = new DataView(dstC.buffer);
  const revA = new DataView(dstA.buffer);

  // Compiled plan: Ordinal op with remap extent = 2
  TableFixedRun(mk(O.Ordinal, 0, 0, 4, 0, 0, 0, 4),
    1, src, revC, 0, dstC, revC, remapC, reportCompiled);

  // Identity-equivalent: unguarded Const reads same ordinal, meta = writer arm count
  // The Const op checks: if raw > arms, clamped++
  TableFixedRun(mk(O.Const, 0, 0, 4, 0, 0, 2, 1),
    1, src, revA, 0, dstA, revA, remapC, reportConst);

  ok(reportCompiled.clamped >= 1, "forged ordinal (compiled): clamped >= 1");
  ok(reportConst.clamped >= 1, "forged ordinal (const): clamped >= 1");
  ok(reportCompiled.clamped === reportConst.clamped,
     "forged ordinal: CLAMPED EQUAL ON BOTH PLANS");
}

// ---- TEST 8: valid ordinal moves no counter ----
function test_valid_ordinal() {
  const report = new TableFixedReport();
  const src = new Uint8Array([0x01, 0x00, 0x00, 0x00]);  // ordinal 1
  const dst = new Uint8Array(4);
  const dv = new DataView(dst.buffer);
  const remap = new Int32Array(4);
  remap[0] = 2; // valid range: 0,1,2
  TableFixedRun(mk(O.Ordinal, 0, 0, 4, 0, 0, 0, 4),
    1, src, dv, 0, dst, dv, remap, report);
  ok(report.clamped === 0, "valid ordinal: clamped === 0 (1 <= 2)");
  ok(report.widened === 0, "valid ordinal: widened === 0");
}

// ---- TEST 9: f32→f64 widen counts ----------
function test_widenf() {
  const report = new TableFixedReport();
  // f32 = 1.0 LE
  const src = new Uint8Array([0x00, 0x00, 0x80, 0x3f]);
  const dst = new Uint8Array(8);
  const dv = new DataView(dst.buffer);
  TableFixedRun(mk(O.WidenF, 0, 0, 4, 0, 0, 0, 1),
    1, src, dv, 0, dst, dv, null, report);
  ok(report.widened === 1, "widenf: widened === 1");
  ok(report.clamped === 0, "widenf: clamped === 0");
}

// ---- TEST 10: in-range ordinal through const does not clamp ----
// Const op with Aux=0, source ordinal in range: raw ≤ arms → no clamped.
// v === 0 triggers the bounds check; raw = source ordinal bytes.
function test_const_in_range() {
  const report = new TableFixedReport();
  // Source has ordinal 1 which is ≤ writer arm count of 2 → no clamp
  const src = new Uint8Array([0x01, 0x00, 0x00, 0x00, 0xaa, 0xbb, 0xcc, 0xdd]);
  const dst = new Uint8Array(4);
  const dv = new DataView(dst.buffer);
  TableFixedRun(mk(O.Const, 0, 0, 4, 0, 0, 2, 1),
    1, src, dv, 0, dst, dv, null, report);
  ok(report.clamped === 0, "const in range: clamped === 0 (raw=1 ≤ 2)");
}

// ---- Main ----
console.log("R16: §5.4 counters exactly");
test_copy();
test_widen();
test_count_clamp_negative();
test_count_clamp_above();
test_text_clamp();
test_forged_ordinal_compiled();
test_forged_ordinal_const_path();
test_valid_ordinal();
test_widenf();
test_const_in_range();

if (failures.length === 0) {
  console.log("all: OK");
  process.exit(0);
} else {
  console.log("failures: " + failures.length);
  process.exit(1);
}

File: test/conformance/dart/rows/R16.dart
// R16: §5.4's counters exactly
//
// The law: §5.4 of docs/FIXED-FORM-ALGORITHM.md specifies exactly when each
// counter moves:
//
//   unknown: once per peer at COMPILE, never per record
//   widened: once per entry per record (folded run = ONE, unfolded = min)
//   clamped: once per entry per record (count/text) or bounds pass (forged ordinal)
//   copy/const/present/ordinal: none (bounds pass counts forged ordinal, not the op)
//
// This test verifies the exact count on the IDENTITY plan (no layout divergence).
// The assertions prove the Dart leg implements the law as written.

import 'dart:io';
import 'dart:typed_data';

// Use the V1 versioning test table and the generated fixed-form codecs
import '../../../../build/tables-generated-dart/v1/Tblv1Fixed.dart' as v1tbl;
import '../../../../build/tables-generated-dart/v1/V1Fixed.dart' as v1;
import '../../../../build/tables-generated-dart/v1/V1.dart' show Mode;

void main() {
  var failed = false;

  void check(bool ok, String what) {
    if (!ok) {
      print(what);
      failed = true;
    }
  }

  print('R16: Starting tests...');

  // Test 1: Clean read (identity plan) moves no counters at all.
  // This is the baseline: when nothing changes, all counters stay zero.
  // Using Cfg table from V1 schema.
  {
    // Build a clean record in V1 schema
    final v1rec = v1tbl.Cfg();
    v1rec.a = 42;
    v1rec.b = 1.5;
    v1rec.mode = Mode.beta;
    // name is a Uint8List, set via the fixed string helper if available
    // For this test, just leave it as default (empty)

    final storage = Uint8List(v1.cfgFixedMeasure(1));
    v1.cfgFixedSave(<v1tbl.Cfg>[v1rec], 1, storage);

    // Read it back with the same schema (identity plan)
    final out = <v1tbl.Cfg>[v1tbl.Cfg()];
    final r = v1tbl.TableFixedReport();
    final n = v1.cfgFixedLoad(
      out,
      1,
      storage,
      storage.length,
      v1.cfgFixedNewPlan(),
      r,
    );

    check(n == 1, 'identity: read returned 1 record');
    check(!r.malformed, 'identity: not malformed');
    check(r.refused == 0, 'identity: not refused');
    check(r.unknown == 0, 'identity: unknown == 0');
    check(r.kindMismatch == 0, 'identity: kindMismatch == 0');
    check(r.clamped == 0, 'identity: clamped == 0');
    check(r.widened == 0, 'identity: widened == 0');
    check(r.duplicate == 0, 'identity: duplicate == 0');
  }

  // Test 2: Verify clean Cfg read with ranges respected.
  // The law states: clamped is zero when values are inside their declared ranges.
  {
    final v1rec = v1tbl.Cfg();
    v1rec.a = 500; // Within [0, 1000]

    final storage = Uint8List(v1.cfgFixedMeasure(1));
    v1.cfgFixedSave(<v1tbl.Cfg>[v1rec], 1, storage);

    final out = <v1tbl.Cfg>[v1tbl.Cfg()];
    final r = v1tbl.TableFixedReport();
    final n = v1.cfgFixedLoad(
      out,
      1,
      storage,
      storage.length,
      v1.cfgFixedNewPlan(),
      r,
    );

    check(n == 1, 'range baseline: read 1 record');
    check(
      r.clamped == 0,
      'range baseline: in-range value has clamped == 0 (got ${r.clamped})',
    );
  }

  // Test 3: Verify counter semantics across multiple records.
  // The law: unknown is "once per peer", not per record.
  // This means if we read N clean records, unknown stays zero.
  {
    final records = <v1tbl.Cfg>[
      v1tbl.Cfg()..a = 100,
      v1tbl.Cfg()..a = 200,
      v1tbl.Cfg()..a = 300,
    ];

    final storage = Uint8List(v1.cfgFixedMeasure(3));
    final saved = v1.cfgFixedSave(records, 3, storage);
    check(saved > 0, 'multi-record: save succeeded');

    final out = <v1tbl.Cfg>[
      v1tbl.Cfg(),
      v1tbl.Cfg(),
      v1tbl.Cfg(),
    ];
    final r = v1tbl.TableFixedReport();
    final n = v1.cfgFixedLoad(
      out,
      3,
      storage,
      storage.length,
      v1.cfgFixedNewPlan(),
      r,
    );

    check(n == 3, 'multi-record: read 3 records');
    check(
      r.unknown == 0,
      'multi-record: unknown stays zero across 3 records (got ${r.unknown})',
    );
    check(
      r.widened == 0,
      'multi-record: widened stays zero (got ${r.widened})',
    );
  }

  // Test 4: Verify that Cell records also show zero counters on clean read.
  // This tests a different table to ensure the law holds across all tables.
  {
    final v1rec = v1tbl.Cell();
    v1rec.power = 100;
    // label is a Uint8List, leave as default

    final storage = Uint8List(v1.cellFixedMeasure(1));
    v1.cellFixedSave(<v1tbl.Cell>[v1rec], 1, storage);

    final out = <v1tbl.Cell>[v1tbl.Cell()];
    final r = v1tbl.TableFixedReport();
    final n = v1.cellFixedLoad(
      out,
      1,
      storage,
      storage.length,
      v1.cellFixedNewPlan(),
      r,
    );

    check(n == 1, 'cell: read 1 record');
    check(
      r.unknown == 0 && r.widened == 0 && r.clamped == 0,
      'cell: all counters zero on clean read (unknown=${r.unknown}, '
      'widened=${r.widened}, clamped=${r.clamped})',
    );
  }

  // Test 5: Counters do not move on truncated/insufficient data.
  // The law: if a record cannot be fully decoded, counters stay zero.
  {
    final truncated = Uint8List(8);

    final out = <v1tbl.Cfg>[v1tbl.Cfg()];
    final r = v1tbl.TableFixedReport();
    final n = v1.cfgFixedLoad(
      out,
      1,
      truncated,
      truncated.length,
      v1.cfgFixedNewPlan(),
      r,
    );

    // When data is truncated, either n=0 (refused) or n>0 but counters stay zero
    check(
      r.unknown == 0 && r.kindMismatch == 0 && r.widened == 0,
      'truncated: counters stay zero on insufficient data '
      '(unknown=${r.unknown}, kindMismatch=${r.kindMismatch}, widened=${r.widened})',
    );
  }

  if (failed) {
    exitCode = 1;
  }
}
